src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp

```diff
@@ -0,0 +1,413 @@
+      const auto& external_bdry_inertial_coords =
+          db::get<domain::Tags::Interface<
+              domain::Tags::BoundaryDirectionsInterior<VolumeDim>,
+              domain::Tags::Coordinates<VolumeDim, Frame::Inertial>>>(box);
+
+      // ------------------------------- (2)
+      // Apply the boundary condition
+      // Loop over external boundaries and set dt_volume_vars on them
+      for (auto& external_direction_and_normals : unit_normal_one_forms) {
+        const auto& direction = external_direction_and_normals.first;
+        const size_t dimension = direction.dimension();
+        const auto& unit_normal_one_form =
+            external_direction_and_normals.second;
+        const size_t slice_grid_points =
+            mesh.extents().slice_away(dimension).product();
+        // Get U on this slice
+        const auto vars =
+            data_on_slice(volume_all_vars, mesh.extents(), dimension,
+                          index_to_slice_at(mesh.extents(), direction));
+        ASSERT(vars.number_of_grid_points() == slice_grid_points,
+               "vars_on_slice has wrong number of grid points.  "
+               "Expected "
+                   << slice_grid_points << ", got "
+                   << vars.number_of_grid_points());
+        // Get dt<U> on this slice
+        const auto dt_vars =
+            data_on_slice(volume_all_dt_vars, mesh.extents(), dimension,
+                          index_to_slice_at(mesh.extents(), direction));
+        // Get characteristic speeds
+        const auto& char_speeds = external_bdry_char_speeds.at(direction);
+        // For external boundaries that are within a horizon,
+        // all characteristic fields are outgoing (toward the singularity)
+        if (BoundaryConditions_detail::min_characteristic_speed<VolumeDim>(
+                char_speeds) >= 0.) {
+          continue;
+        }
+        // Get boundary coordinates
+        const auto& inertial_coords =
+            external_bdry_inertial_coords.at(direction);
+
+        // ------------------------------- (2)
+        // Create a TempTensor that stores all temporaries computed
+        // here and elsewhere
+        TempBuffer<BoundaryConditions_detail::all_local_vars<VolumeDim>> buffer(
+            slice_grid_points);
+        // ------------------------------- (2.2)
+        // Compute local variables, including:
+        // (A) unit normal form to interface
+        // (B) 4metric, inv4metric, lapse, shift on this slice
+        // (C) dampign parameter ConstraintGamma2 on this slice
+        // (D) Compute projection operator on this slice
+        // (E) dt<U> on this slice from `volume_all_dt_vars`
+        BoundaryConditions_detail::local_variables(
+            make_not_null(&buffer), box, direction, dimension, mesh, vars,
+            dt_vars, unit_normal_one_form, char_speeds);
+
+        // FIXME: Are there incoming char speeds on the inner boundary?
```

Comment:
This was intended for debugging only, sorry I should have removed it. The test used here to determine if an interface is inside of a horizon is rudimentary. A proper implementation primarily requires a mechanism to determine all external interfaces in the domain that are completely inside of any horizon. As an inner boundary is likely to remain an inner boundary throughout the evolution, and because classifying all inner boundaries as such is a global operation, this is best done once at the start of the evolution. It can be done either

1. dynamically once after the first time all apparent horizons have been found. or
1. statically by manually tagging all inner interfaces as such during initialization (using perhaps the initial guesses for Ahs).

Going with (2) also adds the complexity of inheriting the "internal boundary" tag appropriately when h-refining the domain. Beyond this, changes needed to add this check to the BC action introduced in this PR will be trivial. There already is a check for boundaries with only outgoing characteristics, this would be an additional if statement checking for the same to be false when the interface is an inner boundary.

Doing either of (1) or (2) requires
- changes to the apparent horizon finding code (where the trigger for tagging inner interfaces must lie), 
- some mechanism to tag 'inner' interfaces as such (perhaps inside the `Interface` tag, this part is not very clear to me yet), and
- additional code to actually do the tagging - whatever be its trigger.

These changes, while small, are best kept separate from this PR as they involve structural changes mostly outside of GH. Especially so since @nilsdeppe is generalizing the whole BC imposition framework, this check might be better placed over there as it is not specific to GH at all. Thoughts?